### PR TITLE
Simplify HIR execution model

### PR DESCRIPTION
**Pull request by havogt**
Advantage of this simplification is a trivial lowering to an IIR (not the one we have today):
just replace the first code snippet (which has vertical regions) by the second snippet. Then all other operations are optimizations of this IR. No extensive "legalization" passes are needed.

### CODE REVIEW

**Comment by havogt** [comment]
Actually from my point of view, this proposal would just be a simple fix for our current model. I believe the better solution would be to discuss "Lukas idea" in more depth (which is roughly the idea Tobias Gysi proposed in the context of his stencil dialect for MLIR).

**Comment by cosunae** [comment]
thanks @havogt, can you summarize can are the changes proposed with respect to current model ? 
What I could see is that vertical_region specifies (opposite to what it does today) a direction. Anything else?
What I didnt understand is your statement that this would simplified the lowering to IIR. The IIR should contain more concepts that are being discussed and we still need to take care of the cases discussed here:
https://github.com/MeteoSwiss-APN/dawn/pull/551
Field versioning, etc would still be required. So what is actually simplifying the lowering in this proposal ? 
Can you also clarify what is Lukas idea?

**Comment by havogt** [comment]
> can you summarize can are the changes proposed with respect to current model ?

Yes, the main change is the explicit direction which solves the problem of deducing solver vs stencils. It's very easy to reason about the behavior. I cannot really summarize the changes, because I don't fully understand the current model.

> The IIR should contain more concepts 

The IIR might contain more concepts, but you don't need to have all the concepts being used after lowering (simplest examples would be the cache attribute). I think it would be advantageous if every SIR concept has a trivial mapping to IIR.

> we still need to take care of the cases discussed here

With the proposed semantics for SIR, I didn't see yet a case that wouldn't work to map to an IIR. Of course I didn't yet specify the IIR, but I also didn't want to propose a complete solution. We should figure this out together if we want to go that route.

> Field versioning, etc would still be required.

No, as far as I am aware, field versioning is not required if you write code compliant with the proposed model. So far, in the discussion, we didn't come up with a case that cannot be handled. But finding these cases is your strength. Please try to break it.

> Can you also clarify what is Lukas idea?

The basic idea is to have pure stencils with explicit k-loops. The pure stencils will make reasoning about data dependencies much easier. The explicit k-loops solve the solver problem, similar to the approach proposed in this PR.

**Comment by twicki** [comment]
### General  
And generally I don't see how this is any different from the initial text? Maybe I'm missing something here?

> If the loop order is forward, any statement executed on a vertical level with index k will be executed after each statement that appears in the current vertical region on every vertical level with indices k-1 and smaller. Each previous vertical region was executed on its full iteration space.

Formally describes a sequential loop as you would translate it.

> Irrespective of the loop order every statement ~~with a horizontal dependency on other statements~~ will be executed after all statements  ~~it depends on~~ _previously appearing in the same vertical region_ are executed on the full horizontal plane in the same vertical level.

Formally describes that there are ij loops around each statement. 

Can you elaborate on where you see the difference?

### Questions

> Yes, the main change is the explicit direction which solves the problem of deducing solver vs stencils. It's very easy to reason about the behavior. I cannot really summarize the changes, because I don't fully understand the current model.

That is in the current model though? SIR has a loop order concept that is either forward or backward. The optimization of making a loop parallel if it is not required to run it sequentially needs to happen somewhere inside dawn where it needs to be understood if the loop at hand is a solver or a stencil.
 
> No, as far as I am aware, field versioning is not required if you write code compliant with the proposed model. So far, in the discussion, we didn't come up with a case that cannot be handled. But finding these cases is your strength. Please try to break it.

This means you would want to explicitly disallow `out = out[i + 1]`?

**Comment by havogt** [comment]
> Can you elaborate on where you see the difference?

If there is no formal difference, perfect. Then the only gain is 80 lines of documentation, that I didn't understand. Maybe it's just a reformulation to my taste, but for me it sounds much simpler.

> somewhere inside dawn where it needs to be understood if the loop at hand is a solver or a stencil.

I don't understand this. Why do we need to understand if it is a solver or a stencil. We "just" need to preserve the semantics. Maybe this is what you mean? But I don't see the need for a binary decision.

> This means you would want to explicitly disallow out = out[i + 1]?

It is disallowed in the proposal.

**Comment by lukasm91** [comment]
>>  This means you would want to explicitly disallow out = out[i + 1]?
>
> It is disallowed in the proposal.

This is the only case I am aware of. I think it's quite obvious that this does not work, because the parfor-specification makes this quite explicit. I would say the main difference is how we approach the problem. We approach the problem by starting with

```
vertical_region(start, end):
    statement1
    statement2
vertical_region(start, end):
    statement3
    statement4
```
corresponds to
```
for k:
    parfor ij:
        statement1
    parfor ij:
        statement2
for k:
    parfor ij:
        statement3
        statement4
```
instead of approaching from current IIR. We would lower by translating the code exactly to the same concept in IIR, and add optimizations (like exchanging loop order + adding caches) afterwards. By this, we would automatically fix the corner cases that we currently observe in the implemented compute model (which we don't fully understand, because we ignore quite some corner cases).

**Comment by lukasm91** [comment]
>> Field versioning, etc would still be required.
> 
> No, as far as I am aware, field versioning is not required if you write code compliant with the proposed model. So far, in the discussion, we didn't come up with a case that cannot be handled. But finding these cases is your strength. Please try to break it.

Not after trivially transforming the code. But when optimizing the code, there might be use cases.
